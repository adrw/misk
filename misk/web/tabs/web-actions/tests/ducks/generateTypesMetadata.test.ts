import HTTPMethod from "http-method-enum"
import {
  generateTypesMetadata,
  ServerTypes,
  TypescriptBaseTypes
} from "../../src/ducks"
import { nonTypedActionAPI, testTypes } from "../testUtilities"

describe("Build typesMetadata from a raw WebActionMetadata", () => {
  it("generateNonTypedGET", () => {
    const typesMetadata = generateTypesMetadata(nonTypedActionAPI)
    expect(typesMetadata.size).toBe(0)
    expect(typesMetadata).toMatchSnapshot()
  })
  it("generateNonTypedPOST", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST
    })
    expect(typesMetadata.size).toBe(1)
    const tmRoot = typesMetadata.get("0")
    expect(tmRoot.serverType).toBe(ServerTypes.JSON)
    expect(tmRoot.typescriptType).toBe(TypescriptBaseTypes.string)
    expect(typesMetadata).toMatchSnapshot()
  })
  it("generateNoNested", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "noRepeatedInt",
      types: testTypes
    })
    expect(typesMetadata.size).toBe(2)
    expect(typesMetadata).toMatchSnapshot()
  })
  it("generateNested", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "nestedNoRepeatedInt",
      types: testTypes
    })
    expect(typesMetadata.size).toBe(3)
    const tmRoot = typesMetadata.get("0")
    const tmFieldGroup = typesMetadata.get(tmRoot.idChildren.first())
    expect(tmFieldGroup.serverType).toBe("noRepeatedInt")
    expect(tmFieldGroup.typescriptType).toBeNull()
    const tmField = typesMetadata.get(tmFieldGroup.idChildren.first())
    expect(tmField.serverType).toBe(ServerTypes.Int)
    expect(tmField.typescriptType).toBe(TypescriptBaseTypes.number)
    expect(typesMetadata).toMatchSnapshot()
  })
  it("generateRepeated", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "repeatedDouble",
      types: testTypes
    })
    expect(typesMetadata.size).toBe(3)
    const tmRoot = typesMetadata.get("0")
    expect(tmRoot.idChildren.size).toBe(1)
    const tmParent = typesMetadata.get(tmRoot.idChildren.first())
    expect(tmParent.repeated).toBe(true)
    expect(tmParent.serverType).toBe(ServerTypes.Double)
    expect(tmParent.typescriptType).toBeNull()
    const tmChild = typesMetadata.get(tmParent.idChildren.first())
    expect(tmChild.repeated).toBe(false)
    expect(tmChild.serverType).toBe(ServerTypes.Double)
    expect(tmChild.typescriptType).toBe(TypescriptBaseTypes.number)
    expect(typesMetadata).toMatchSnapshot()
  })
  it("generateRepeatedNested", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "repeatedNestedNoRepeatedInt",
      types: testTypes
    })
    expect(typesMetadata.size).toBe(5)
    const tmRoot = typesMetadata.get("0")
    expect(tmRoot.idChildren.size).toBe(1)
    const tmParent = typesMetadata.get(tmRoot.idChildren.first())
    expect(tmParent.repeated).toBe(true)
    expect(tmParent.serverType).toBe("nestedNoRepeatedInt")
    expect(tmParent.typescriptType).toBeNull()
    const tmChild = typesMetadata.get(tmParent.idChildren.first())
    expect(tmChild.repeated).toBe(false)
    expect(tmChild.serverType).toBe("nestedNoRepeatedInt")
    expect(tmChild.typescriptType).toBeNull()
    const tmFieldGroup = typesMetadata.get(tmChild.idChildren.first())
    expect(tmFieldGroup.repeated).toBe(false)
    expect(tmFieldGroup.serverType).toBe("noRepeatedInt")
    expect(tmFieldGroup.typescriptType).toBeNull()
    const tmField = typesMetadata.get(tmFieldGroup.idChildren.first())
    expect(tmField.repeated).toBe(false)
    expect(tmField.serverType).toBe(ServerTypes.Int)
    expect(tmField.typescriptType).toBe(TypescriptBaseTypes.number)
    expect(typesMetadata).toMatchSnapshot()
  })
  it("generateRepeatedNestedRepeated", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "repeatedNestedRepeatedDouble",
      types: testTypes
    })
    expect(typesMetadata.size).toBe(6)
    const tmRoot = typesMetadata.get("0")
    expect(tmRoot.idChildren.size).toBe(1)
    const tmParent = typesMetadata.get(tmRoot.idChildren.first())
    expect(tmParent.repeated).toBe(true)
    expect(tmParent.serverType).toBe("nestedRepeatedDouble")
    expect(tmParent.typescriptType).toBeNull()
    const tmChild = typesMetadata.get(tmParent.idChildren.first())
    expect(tmChild.repeated).toBe(false)
    expect(tmChild.serverType).toBe("nestedRepeatedDouble")
    expect(tmChild.typescriptType).toBeNull()
    const tmFieldGroup = typesMetadata.get(tmChild.idChildren.first())
    expect(tmFieldGroup.repeated).toBe(false)
    expect(tmFieldGroup.serverType).toBe("repeatedDouble")
    expect(tmFieldGroup.typescriptType).toBeNull()
    const tmFieldParent = typesMetadata.get(tmFieldGroup.idChildren.first())
    expect(tmFieldParent.repeated).toBe(true)
    expect(tmFieldParent.serverType).toBe(ServerTypes.Double)
    expect(tmFieldParent.typescriptType).toBeNull()
    const tmFieldChild = typesMetadata.get(tmFieldParent.idChildren.first())
    expect(tmFieldChild.repeated).toBe(false)
    expect(tmFieldChild.serverType).toBe(ServerTypes.Double)
    expect(tmFieldChild.typescriptType).toBe(TypescriptBaseTypes.number)
    expect(typesMetadata).toMatchSnapshot()
  })
  it("generate multiple top level fields", () => {
    const typesMetadata = generateTypesMetadata({
      ...nonTypedActionAPI,
      dispatchMechanism: HTTPMethod.POST,
      requestType: "multipleFlatFields",
      types: testTypes
    })
    expect(typesMetadata.size).toBe(5)
    expect(typesMetadata).toMatchSnapshot()
  })
})
