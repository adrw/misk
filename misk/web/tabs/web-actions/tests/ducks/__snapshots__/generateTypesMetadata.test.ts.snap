// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Build typesMetadata from a raw WebActionMetadata generate multiple top level fields 1`] = `
Immutable.OrderedMap {
  "0": Object {
    "id": "0",
    "idChildren": Immutable.OrderedSet [
      "15",
      "16",
      "17",
      "18",
    ],
    "idParent": "0",
    "name": "",
    "repeated": false,
    "serverType": null,
    "typescriptType": null,
  },
  "15": Object {
    "id": "15",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "0",
    "name": "Field 1",
    "repeated": false,
    "serverType": "Double",
    "typescriptType": "number",
  },
  "16": Object {
    "id": "16",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "0",
    "name": "Field 2",
    "repeated": false,
    "serverType": "Int",
    "typescriptType": "number",
  },
  "17": Object {
    "id": "17",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "0",
    "name": "Field 3",
    "repeated": false,
    "serverType": "Long",
    "typescriptType": "number",
  },
  "18": Object {
    "id": "18",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "0",
    "name": "Field 4",
    "repeated": false,
    "serverType": "Short",
    "typescriptType": "number",
  },
}
`;

exports[`Build typesMetadata from a raw WebActionMetadata generateNested 1`] = `
Immutable.OrderedMap {
  "0": Object {
    "id": "0",
    "idChildren": Immutable.OrderedSet [
      "2",
    ],
    "idParent": "0",
    "name": "",
    "repeated": false,
    "serverType": null,
    "typescriptType": null,
  },
  "2": Object {
    "id": "2",
    "idChildren": Immutable.OrderedSet [
      "3",
    ],
    "idParent": "0",
    "name": "Nested Int Field",
    "repeated": false,
    "serverType": "noRepeatedInt",
    "typescriptType": null,
  },
  "3": Object {
    "id": "3",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "2",
    "name": "Int Field",
    "repeated": false,
    "serverType": "Int",
    "typescriptType": "number",
  },
}
`;

exports[`Build typesMetadata from a raw WebActionMetadata generateNoNested 1`] = `
Immutable.OrderedMap {
  "0": Object {
    "id": "0",
    "idChildren": Immutable.OrderedSet [
      "1",
    ],
    "idParent": "0",
    "name": "",
    "repeated": false,
    "serverType": null,
    "typescriptType": null,
  },
  "1": Object {
    "id": "1",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "0",
    "name": "Int Field",
    "repeated": false,
    "serverType": "Int",
    "typescriptType": "number",
  },
}
`;

exports[`Build typesMetadata from a raw WebActionMetadata generateNonTypedGET 1`] = `Immutable.OrderedMap {}`;

exports[`Build typesMetadata from a raw WebActionMetadata generateNonTypedPOST 1`] = `
Immutable.OrderedMap {
  "0": Object {
    "id": "0",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "",
    "name": "",
    "repeated": false,
    "serverType": "JSON",
    "typescriptType": "string",
  },
}
`;

exports[`Build typesMetadata from a raw WebActionMetadata generateRepeated 1`] = `
Immutable.OrderedMap {
  "0": Object {
    "id": "0",
    "idChildren": Immutable.OrderedSet [
      "4",
    ],
    "idParent": "0",
    "name": "",
    "repeated": false,
    "serverType": null,
    "typescriptType": null,
  },
  "4": Object {
    "id": "4",
    "idChildren": Immutable.OrderedSet [
      "5",
    ],
    "idParent": "0",
    "name": "Repeated Double Field",
    "repeated": true,
    "serverType": "Double",
    "typescriptType": null,
  },
  "5": Object {
    "id": "5",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "4",
    "name": "Repeated Double Field",
    "repeated": false,
    "serverType": "Double",
    "typescriptType": "number",
  },
}
`;

exports[`Build typesMetadata from a raw WebActionMetadata generateRepeatedNested 1`] = `
Immutable.OrderedMap {
  "0": Object {
    "id": "0",
    "idChildren": Immutable.OrderedSet [
      "6",
    ],
    "idParent": "0",
    "name": "",
    "repeated": false,
    "serverType": null,
    "typescriptType": null,
  },
  "6": Object {
    "id": "6",
    "idChildren": Immutable.OrderedSet [
      "7",
    ],
    "idParent": "0",
    "name": "Repeated Nested Int Field",
    "repeated": true,
    "serverType": "nestedNoRepeatedInt",
    "typescriptType": null,
  },
  "7": Object {
    "id": "7",
    "idChildren": Immutable.OrderedSet [
      "8",
    ],
    "idParent": "6",
    "name": "Repeated Nested Int Field",
    "repeated": false,
    "serverType": "nestedNoRepeatedInt",
    "typescriptType": null,
  },
  "8": Object {
    "id": "8",
    "idChildren": Immutable.OrderedSet [
      "9",
    ],
    "idParent": "7",
    "name": "Nested Int Field",
    "repeated": false,
    "serverType": "noRepeatedInt",
    "typescriptType": null,
  },
  "9": Object {
    "id": "9",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "8",
    "name": "Int Field",
    "repeated": false,
    "serverType": "Int",
    "typescriptType": "number",
  },
}
`;

exports[`Build typesMetadata from a raw WebActionMetadata generateRepeatedNestedRepeated 1`] = `
Immutable.OrderedMap {
  "0": Object {
    "id": "0",
    "idChildren": Immutable.OrderedSet [
      "10",
    ],
    "idParent": "0",
    "name": "",
    "repeated": false,
    "serverType": null,
    "typescriptType": null,
  },
  "10": Object {
    "id": "10",
    "idChildren": Immutable.OrderedSet [
      "11",
    ],
    "idParent": "0",
    "name": "Nested Repeated Double Field",
    "repeated": true,
    "serverType": "nestedRepeatedDouble",
    "typescriptType": null,
  },
  "11": Object {
    "id": "11",
    "idChildren": Immutable.OrderedSet [
      "12",
    ],
    "idParent": "10",
    "name": "Nested Repeated Double Field",
    "repeated": false,
    "serverType": "nestedRepeatedDouble",
    "typescriptType": null,
  },
  "12": Object {
    "id": "12",
    "idChildren": Immutable.OrderedSet [
      "13",
    ],
    "idParent": "11",
    "name": "Nested Double Field",
    "repeated": false,
    "serverType": "repeatedDouble",
    "typescriptType": null,
  },
  "13": Object {
    "id": "13",
    "idChildren": Immutable.OrderedSet [
      "14",
    ],
    "idParent": "12",
    "name": "Repeated Double Field",
    "repeated": true,
    "serverType": "Double",
    "typescriptType": null,
  },
  "14": Object {
    "id": "14",
    "idChildren": Immutable.OrderedSet [],
    "idParent": "13",
    "name": "Repeated Double Field",
    "repeated": false,
    "serverType": "Double",
    "typescriptType": "number",
  },
}
`;
